### PR TITLE
Card Management System v1.3

### LINKED ISSUE #19: As a creator I can view all cards in a lesson
So I can at a glance find out what cards are there in the lesson, and proceed to add (#50), delete (#51) or modify them (#52 #53 #54)

### LINKED ISSUE #50: As a creator I can add a card to a lesson
So I can update a lesson to test more and new content

### LINKED ISSUE #51: As a creator I can delete a card from a lesson
So I can remove outdated, incorrect or irrelevant cards

### LINKED ISSUE #110: Showing incorrect message for deleting non-existent lesson
Currently, if I try to delete non-existent lesson, it will feedback that it is an invalid command instead of an invalid lesson or something else.

**Recreate by** 
```
listLesson # index till 2
deleteLesson 3 # invalid message
```

**Expected outcome**
```
listLesson # index till 2
deleteLesson 3 # invalid lesson index or no such lesson
```

### LINKED ISSUE #112: After opening and closing a lesson, deleting the lesson causes program failure
Currently, if I try to delete a lesson after opening and closing it, the program fails.

**Steps to replicate error** 
```
addLesson n/Capitals of the world c/Country c/Capital o/Hint
openLesson 1
closeLesson 1
deleteLesson 1
```

A `NullPointerException` is thrown after the last command.

### LINKED ISSUE #114: Showing incorrect message for opening non-existent lesson
If the user enters an `openLesson [INDEX]` command in the correct format, in an attempt to delete a non-existent lesson, the feedback provided will state that it is an invalid command format. This should not be the case because the format is correct. Instead, the user should be notified that the index provided is invalid.

Assuming there is only 1 lesson at (index 1) in memory:

```
deleteLesson 3
```

### LINKED ISSUE #115: As a user, my lessons are automatically saved to the hard disk.
So that changes made to lessons previously (add/delete/modify) are still available in subsequent launches of BrainTrain.

### PR DESCRIPTION
**Pull request by DigiPie**
[Work-in-progress]
### Implement Card-aspect of Card Management System (CMS)
This is a follow up to the PR: **Card Management System mid-v1.3 #107** which was merged in milestone mid-v1.3.

#### Enhance CMS-related commands
- Integrates `StorageManager` with `AddLessonCommand`, `DeleteLessonCommand` and `CloseLessonCommand`. Changes to the lessons loaded in memory are automatically saved to the hard disk (closes #115).
- Implement `listCards` command - use it to list the opened lesson's cards (closes #19).
- Implement `addCard` command - use it to add card to the opened lesson (closes #50).
- Implement `deleteCard` command - use it to delete card from the opened lesson (closes #51).

#### Refactoring
- Updates `LessonList` with `getOpenedLessonCards()`, `getOpenedLessonCoreHeaders`, `getOpenedLessonOptionalHeaders` methods to conform to the *Law of Demeter*. For example, `getOpenedLesson().getCards()` is refactored to `getOpenedLessonCards()`.

#### Bug fixes & testing
- Fixes bug in `deleteLesson` command - program no longer crashes (fixes #112)
- Fixes bugs in `openLesson` and `deleteLesson` command - program now shows more appropriate feedback message to users (fixes #114 and fixes #110)
- Updates `ManagementCommandsIntegrationTest ` to test for bug #112 in case of regression
- Updates `LessonList` related-testing classes are as follows:
   - Adds `LessonListBuilder`: Constructs `Lesson` objects  for `TypicalLessons` to use
   - Adds `TypicalLessonLists`: Contains sample `Lesson` objects for `LessonTest` to use
   - Updates `LessonListTest`: Uses `LessonListBuilder` and `TypicalLessonLists` for better code quality
- Enhance `ManagementModelManagerTest`, `LessonTest` and `LessonList` tests to achieve 💯 coverage.
- Enhance tests for commands under `seedu.address.logic.commands.management` package to achieve 💯 coverage.

### Documentation
- Update `UserGuide.adoc` with management command usage.

### CODE REVIEW

File: src/test/java/seedu/address/logic/LogicManagerTest.java [hunk 1]
```diff
@@ -317,4 +326,21 @@ private void assertHistoryCorrect(String... expectedCommands) {
...
+    @Test
+    public void testManagementInputOutput() {
+        Lesson validLesson = new LessonBuilder().withNoOptionalHeaders().build();
+        String command = AddLessonCommand.COMMAND_WORD + " "
+                + PREFIX_LESSON_NAME + DEFAULT_NAME + " "
+                + PREFIX_CORE_QUESTION + DEFAULT_CORE_HEADER_1 + " "
+                + PREFIX_CORE_ANSWER + DEFAULT_CORE_HEADER_2;
+
+        try {
+            logic.execute(command);
```
Comment (halcyoneee) [inline src/test/java/seedu/address/logic/LogicManagerTest.java]: As there is a command test util method in `LogicManagerTest`, consider using `assertCommandSuccess` method instead which tests it more comprehensively.

File: docs/UserGuide.adoc [hunk 2]
```diff
@@ -90,26 +98,31 @@ Examples:
...
-The lesson can have any number of `OPTIONAL` headers (including 0).
-
-The `OPTIONAL` headers are optional. The values of these headers are displayed during quiz mode when you enter <<hint, `\hint`>>.
+`OPTIONAL` values are displayed during quiz mode when you enter <<hint, `\hint`>>.
 
 Examples:
 
-* `addLesson n/MedicineTrivia c/Proposition c/TrueOrFalse`
-* `addLesson n/Country Capitals c/Country c/Capital o/First letter of capital`
+* `addLesson n/Upper limb anatomy q/Statement a/TrueOrFalse`
```
[the remark this answers is not in the document]
Comment (DigiPie) [reply docs/UserGuide.adoc]: Noted on this. I will provide a screenshot of what a user will see after adding a lesson once I am done with UI by mid-v1.4.

File: docs/UserGuide.adoc [hunk 3]
```diff
@@ -127,6 +140,7 @@ Examples:
 * `deleteLesson 2` +
```
[the remark this answers is not in the document]
Comment (DigiPie) [reply docs/UserGuide.adoc]: I have resolved this, please view my latest `UserGuide.adoc`.

File: src/test/java/seedu/address/logic/LogicManagerTest.java (hunk 1 above)
Comment (DigiPie) [reply src/test/java/seedu/address/logic/LogicManagerTest.java]: Oh alright. I will fix this by mid-v1.4.

File: src/main/java/seedu/address/logic/commands/management/AddCardCommand.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        ManagementModel mgtModel = requireManagementModel(model);
+
+        try {
+            mgtModel.addCardToOpenedLesson(toAdd);
+            return new CommandResult(String.format(MESSAGE_SUCCESS, toAdd));
+        } catch (IllegalArgumentException e) {
+            throw new CommandException(
+                    String.format(MESSAGE_INVALID_CORE_COUNT,
+                            mgtModel.getOpenedLessonCoreHeaders().size(),
+                            mgtModel), e);
```
[the remark this answers is not in the document]
Comment (DigiPie) [reply src/main/java/seedu/address/logic/commands/management/AddCardCommand.java]: Will fix this in v1.4!

File: docs/UserGuide.adoc [hunk 5]
```diff
@@ -127,13 +121,18 @@ Examples:
...
 
 Usage rules:
 
-* Deletes the lesson at the specified `INDEX` of the numbered lesson list.
 * To view the list, enter the command <<listLessons, `listLessons`>>.
+* You must specify the `INDEX` of the lesson in the numbered lesson list.
+
+[NOTE]
+====
+You have to open a lesson before you can use the following commands: <<listCards, `listCards`>>, <<addCard, `addCard`>> and <<deleteCard, `deleteCard`>> commands.
```
Comment (halcyoneee) [inline docs/UserGuide.adoc]: Seems a bit unrelated since this is deleteLesson but you are telling me about cards

**Pull request merged**